account_mtd/models/mtd_hello_world.py

```diff
@@ -0,0 +1,456 @@
+# -*- coding: utf-8 -*-
+
+import requests
+import json
+import logging
+import werkzeug
+import urllib
+
+from odoo import models, fields, api, exceptions
+from datetime import datetime, timedelta
+from urllib.parse import urlparse
+
+
+class MtdHelloWorld(models.Model):
+    _name = 'mtd.hello_world'
+    _description = "Hello world to test connection between Odoo and HMRC"
+    
+    _logger = logging.getLogger(__name__)
+    
+    name = fields.Char('Name', required=True, readonly=True)
+    api_name = fields.Many2one(comodel_name="mtd.api", required=True)
+    hmrc_configuration = fields.Many2one(comodel_name="mtd.hmrc_configuration", string='HMRC Configuration')
+    scope = fields.Char(related="api_name.scope")
+    response_from_hmrc = fields.Text(string="Response From HMRC", readonly=True)
+    which_button_type_clicked = fields.Char(string="which_button")
+    path = fields.Char(string="sandbox_url")
+    current_record = fields.Char()
+    # need a variable which will keep the record id so that when we
+    
+    @api.multi
+    def action_hello_world_connection(self):
+        if not self.hmrc_configuration:
+            raise exceptions.Warning("Please select HMRC configuration before continuing!")
+        # making sure that the endpoint is what we have used in the data file as the name can be changed anytime.
+        endpoint_record = self.env['ir.model.data'].search([
+            ('res_id', '=', self.id), ('model', '=', 'mtd.hello_world')
+        ])
+
+        if endpoint_record.name == "mtd_hello_world_endpoint":
```

Comment:
I think there is a more pythonic way of doing this, I will sit at your desk to run through it

Reply:
In fact, is endpoint_record something which a user can edit? in which case it is probably better to have it hard-coded like this.

Reply:
```
mapping = {
    'mtd_hello_world_endpoint': '_handle_mtd_hello_world_endpoint',
    'mtd_hello_application_endpoint': '_handle_mtd_hello_application_endpoint',
    'mtd_hello_user_endpoint': '_handle_mtd_hello_user_endpoint'
}
if not mapping.get(endpoint_record.name):
    raise exceptions.Warning("Could not connect to HMRC! \nThis is not a valid HMRC service connection")
return getattr(self, mapping.get(endpoint_record.name))
```
I think should work

Reply:
Yep another step in the right direction (except you've not called the method returned by getattr()), but why not:

```
request_handlers = {
   ...
}
try:
    handler_name = request_handlers[endpoint_record.name]
except KeyError:
    raise exceptions.Warning(...)
else:
    handle_request = getattr(self, handler_name)
    return handle_request()
```

or

```
handler_name = request_handlers.get(endpoint_record.name)
if handler_name:
    handle_request = getattr(self, handler_name)
    return handle_request()
else:
    raise exceptions.Warning(...)
```


It's also possible to define handler_name in the above thus:

```
handler_name = "_handle_" + endpoint_record.name
try:
    handle_request = getattr(self, handler_name)
except AttributeError:
    raise exceptions.Warning()
else:
    return handle_request()
```

BUT...

I think this actually fits the strategy pattern [1] but I don't know where the split between classes (or models in Odoo's case) would go without a much better understanding of how all these bits of logic and data are interacting.

[1] https://en.wikipedia.org/wiki/Strategy_pattern